File: EasyMSXPython/BrokerStrategies.py
# BrokerStrategies.py

import blpapi
from BrokerStrategy import BrokerStrategy

GET_BROKER_STRATEGIES = blpapi.Name("GetBrokerStrategiesWithAssetClass")
ERROR_INFO = blpapi.Name("ErrorInfo")

class BrokerStrategies:
    
    def __init__(self,broker):
        self.broker = broker
        self.strategies=[]
        self.loadStrategies()
        
    def __iter__(self):
        return self.strategies.__iter__()

    def loadStrategies(self):
        request = self.broker.parent.easyMSX.emsxService.createRequest(str(GET_BROKER_STRATEGIES))
        request.set("EMSX_BROKER", self.broker.name)
        request.set("EMSX_ASSET_CLASS", self.broker.assetClass)
        self.broker.parent.easyMSX.submitRequest(request, self.processMessage)
        
    def processMessage(self, msg):
        
        if msg.messageType() == ERROR_INFO:
            errorCode = msg.getElementAsInteger("ERROR_CODE")
            errorMessage = msg.getElementAsString("ERROR_MESSAGE")
            print "GetTeams >> ERROR CODE: %d\tERROR MESSAGE: %s" % (errorCode,errorMessage)

        elif msg.messageType() == GET_BROKER_STRATEGIES:
            
            strats = msg.getElement("EMSX_STRATEGIES")
            
            for s in strats.values():
                if len(s) > 0:
                    self.strategies.append(BrokerStrategy(self,s))
                                  
__copyright__ = """
Copyright 2017. Bloomberg Finance L.P.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to
deal in the Software without restriction, including without limitation the
rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
sell copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:  The above
copyright notice and this permission notice shall be included in all copies
or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
IN THE SOFTWARE.
"""

File: EasyMSXCSharp/EasyMSX/BrokerStrategies.cs
using System.Collections.Generic;
using Name = Bloomberglp.Blpapi.Name;
using LogLevels = com.bloomberg.emsx.samples.Log.LogLevels;
using Request = Bloomberglp.Blpapi.Request;
using Message = Bloomberglp.Blpapi.Message;
using Element = Bloomberglp.Blpapi.Element;
using System.Collections;

namespace com.bloomberg.emsx.samples {

    public class BrokerStrategies: IEnumerable<BrokerStrategy> {
	
        private static readonly Name 	GET_BROKER_STRATEGIES = new Name("GetBrokerStrategiesWithAssetClass");
	    private static readonly Name 	ERROR_INFO = new Name("ErrorInfo");
	
	    private List<BrokerStrategy> strategies = new List<BrokerStrategy>();
	    internal Broker broker;
	
	    internal BrokerStrategies(Broker broker) {
		    this.broker = broker;
		    loadStrategies();
	    }

	    private void loadStrategies() {
		    Log.LogMessage(LogLevels.DETAILED, "Broker [" + broker.name + "]: Loading Strategies");
		    Request request = broker.parent.emsxapi.emsxService.CreateRequest(GET_BROKER_STRATEGIES.ToString());
    	    request.Set("EMSX_BROKER", broker.name);
    	    request.Set("EMSX_ASSET_CLASS", broker.assetClass.ToString());
		    broker.parent.emsxapi.submitRequest(request,new BrokerStrategiesHandler(this));		
	    }
	
	    class BrokerStrategiesHandler : MessageHandler {
		
		    BrokerStrategies brokerStrategies;
		
		    internal BrokerStrategiesHandler(BrokerStrategies brokerStrategies) {
			    this.brokerStrategies = brokerStrategies;
		    }

		    public void processMessage(Message message) {
			
			    Log.LogMessage(LogLevels.DETAILED, "Broker Strategies ["+ brokerStrategies.broker.name + "]: processing message");
			
	    	    if(message.MessageType.Equals(ERROR_INFO)) {
	        	    Log.LogMessage(LogLevels.BASIC, "Broker Strategies ["+ brokerStrategies.broker.name + "]: processing RESPONSE error");
	    		    int errorCode = message.GetElementAsInt32("ERROR_CODE");
	    		    string errorMessage = message.GetElementAsString("ERROR_MESSAGE");
	    		    Log.LogMessage(LogLevels.BASIC, "Broker Strategies ["+ brokerStrategies.broker.name + "]: [" + errorCode + "] " + errorMessage);
	    	    } else if(message.MessageType.Equals(GET_BROKER_STRATEGIES)) {
	        	    Log.LogMessage(LogLevels.DETAILED, "Broker Strategies ["+ brokerStrategies.broker.name + "]: processing succesful RESPONSE");
	    		
	    		    Element strategies = message.GetElement("EMSX_STRATEGIES");
	    		
				    int numValues = strategies.NumValues;
	    		
	    		    for(int i = 0; i < numValues; i++) {
	    			
	    			    string strategy = strategies.GetValueAsString(i);
	    			    if(strategy.Length>0) {
	    				    BrokerStrategy newBrokerStrategy = new BrokerStrategy(brokerStrategies, strategy);
	    				    brokerStrategies.add(newBrokerStrategy);
	    	        	    Log.LogMessage(LogLevels.DETAILED, "Broker Strategies ["+ brokerStrategies.broker.name + "]: added new strategy " + newBrokerStrategy.name);
	    			    }
	    		    }
	    	    }		
	        }
	    }

        public IEnumerator<BrokerStrategy> GetEnumerator()
        {
            return strategies.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return this.GetEnumerator();
        }

	    public BrokerStrategy get(int index) {
		    return strategies[index];
	    }
	
	    public BrokerStrategy get(string name) {
		    foreach(BrokerStrategy s in strategies){
			    if(s.name.Equals(name))
			     return s;
		    }
		    return null;
	    }
	
	    public void add(BrokerStrategy newBrokerStrategy) {
		    strategies.Add(newBrokerStrategy);
	    }
    }
}